### ISSUE TITLE
Inherited user properties show source as "local" in "zfs get"

### ISSUE THREAD

**Issue by rblank** [zfsonlinux/zfs#4099]
Using zfs-0.6.5.3 on Gentoo. When setting a user property on a filesystem, it is correctly inherited by a child filesystem. But the source of the property is displayed as "local" on the child, even though it wasn't overridden. This makes it impossible to determine if a property is inherited by the child or overridden. Filtering by source "inherited" results in an empty set, whereas "local" shows the inherited property.

```
$ zfs create pool/store/test
$ zfs create pool/store/test/sub
$ zfs set user:prop=parent pool/store/test
$ zfs get user:prop pool/store/test
NAME             PROPERTY   VALUE      SOURCE
pool/store/test  user:prop  parent     local
$ zfs get user:prop pool/store/test/sub
NAME                 PROPERTY   VALUE      SOURCE
pool/store/test/sub  user:prop  parent     local
$ zfs get -s inherited user:prop pool/store/test/sub
$ zfs get -s local user:prop pool/store/test/sub
NAME                 PROPERTY   VALUE      SOURCE
pool/store/test/sub  user:prop  parent     local
```

**Comment by ilovezfs** [comment]
I cannot reproduce this issue with 0.6.5.3 or 0.6.5.3-r1.

**Comment by rblank** [comment]
Interesting. I'm actually running 0.6.5.3-r1, too, but I have other machines still on 0.6.5.3.

tl;dr: This is an issue with 0.6.5.3-r1 only. I can't reproduce the issue with 0.6.5.3. So I'm going to open a bug with Gentoo. IIUC -r1 is more than just an ebuild change, it's an actual patchset against zfs.

I use the following script:
```shell
#!/bin/bash
set -ev
dd if=/dev/zero of=test-pool.vol bs=1M count=64
zpool create -R $PWD test-pool $PWD/test-pool.vol
cd test-pool
zfs create test-pool/test
zfs create test-pool/test/sub
zfs set user:prop=parent test-pool/test
zfs get user:prop test-pool/test
zfs get user:prop test-pool/test/sub
zfs get -s inherited user:prop test-pool/test/sub
zfs get -s local user:prop test-pool/test/sub
cd ..
zpool export test-pool
```
Here's a test run on 0.6.5.3-r1. Note that the user property is always shown as "local".
```
dd if=/dev/zero of=test-pool.vol bs=1M count=64
64+0 records in
64+0 records out
67108864 bytes (67 MB) copied, 0.0663682 s, 1.0 GB/s
zpool create -R $PWD test-pool $PWD/test-pool.vol
cd test-pool
zfs create test-pool/test
zfs create test-pool/test/sub
zfs set user:prop=parent test-pool/test
zfs get user:prop test-pool/test
NAME            PROPERTY   VALUE      SOURCE
test-pool/test  user:prop  parent     local
zfs get user:prop test-pool/test/sub
NAME                PROPERTY   VALUE      SOURCE
test-pool/test/sub  user:prop  parent     local
zfs get -s inherited user:prop test-pool/test/sub
zfs get -s local user:prop test-pool/test/sub
NAME                PROPERTY   VALUE      SOURCE
test-pool/test/sub  user:prop  parent     local
cd ..
zpool export test-pool
```
And here's a run on 0.6.5.3. This time it's shown properly as "inherited".
```
dd if=/dev/zero of=test-pool.vol bs=1M count=64
64+0 records in
64+0 records out
67108864 bytes (67 MB) copied, 0.100981 s, 665 MB/s
zpool create -R $PWD test-pool $PWD/test-pool.vol
cd test-pool
zfs create test-pool/test
zfs create test-pool/test/sub
zfs set user:prop=parent test-pool/test
zfs get user:prop test-pool/test
NAME            PROPERTY   VALUE      SOURCE
test-pool/test  user:prop  parent     local
zfs get user:prop test-pool/test/sub
NAME                PROPERTY   VALUE      SOURCE
test-pool/test/sub  user:prop  parent     inherited from test-pool/test
zfs get -s inherited user:prop test-pool/test/sub
NAME                PROPERTY   VALUE      SOURCE
test-pool/test/sub  user:prop  parent     inherited from test-pool/test
zfs get -s local user:prop test-pool/test/sub
cd ..
zpool export test-pool
```

**Comment by rblank** [comment]
Gentoo bug is: https://bugs.gentoo.org/show_bug.cgi?id=568270

**Comment by ilovezfs** [comment]
Ah yeah, I can reproduce it 0.6.5.3-r1. It's order dependent.

```
openzfs ~ # zfs create o3xpool/foo2
openzfs ~ # zfs create o3xpool/foo2/bar2
openzfs ~ # zfs set myuser:myprop=myparent o3xpool/foo2
openzfs ~ # zfs get myuser:myprop o3xpool/foo2
NAME          PROPERTY       VALUE          SOURCE
o3xpool/foo2  myuser:myprop  myparent       local
openzfs ~ # zfs get myuser:myprop o3xpool/foo2/bar2
NAME               PROPERTY       VALUE          SOURCE
o3xpool/foo2/bar2  myuser:myprop  myparent       local
openzfs ~ # zfs create o3xpool/foo2/bar3
openzfs ~ # zfs get myuser:myprop o3xpool/foo2/bar3
NAME               PROPERTY       VALUE          SOURCE
o3xpool/foo2/bar3  myuser:myprop  myparent       inherited from o3xpool/foo2
```
